theano/gpuarray/tests/test_dnn.py

```diff
@@ -2298,3 +2298,215 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+    # images is generated in the expected (NCHW) format
+    img_dims = (5, 3, 16, 16)
+    img = np.random.randint(low=0, high=256, size=img_dims).astype(theano.config.floatX)
+    scale_height = 0.25
+    scale_width = 0.75
+
+    # Transformation matrix
+    transform = [[-1, 0, 0],
+                 [0, -1, 0]]
+    theta = np.asarray(img_dims[0] * [transform], dtype=theano.config.floatX)
+
+    # Create symbolic variables for inputs and transformations
+    t_img = T.tensor4('img')
+    t_theta = T.tensor3('theta')
+
+    st_dnn = dnn.dnn_spatialtf(t_img, t_theta,
+                               scale_height=scale_height,
+                               scale_width=scale_width)
+    st_dnn_func = theano.function([t_img, t_theta], st_dnn)
+    # Check if function graph contains the spatial transformer's grid and sampler Ops
+    apply_nodes = st_dnn_func.maker.fgraph.apply_nodes
+    assert any([isinstance(node.op, dnn.GpuDnnTransformerGrid) for node in apply_nodes])
+    assert any([isinstance(node.op, dnn.GpuDnnTransformerSampler) for node in apply_nodes])
+
+    img_out_gpu = st_dnn_func(img, theta)
+    img_out_gpu = np.asarray(img_out_gpu)
+
+    # Setup CPU Op
+    st_cpu = spatialtf_cpu(t_img, t_theta, scale_height, scale_width, 'nearest')
+    st_cpu_func = theano.function([t_img, t_theta], st_cpu, mode=mode_without_gpu)
+    img_out_cpu = st_cpu_func(img, theta)
+
+    utt.assert_allclose(img_out_cpu, img_out_gpu, rtol=1e-2, atol=1e-2)
+
+
+def test_dnn_spatialtf_grad():
+    if not dnn.dnn_available(test_ctx_name):
+        raise SkipTest(dnn.dnn_available.msg)
+
+    utt.seed_rng()
+
+    inputs = T.tensor4('inputs')
+    theta = T.tensor3('theta')
+
+    out = dnn.dnn_spatialtf(inputs, theta, scale_height=0.25, scale_width=0.75)
+    out_mean = T.mean(out)
+    mean_gi = T.grad(out_mean, [inputs])
+    mean_gt = T.grad(out_mean, [theta])
+
+    f_gi = theano.function([inputs, theta], mean_gi)
+    assert any([isinstance(node.op, dnn.GpuDnnTransformerGradI)
+                for node in f_gi.maker.fgraph.apply_nodes])
+
+    f_gt = theano.function([inputs, theta], mean_gt)
+    assert any([isinstance(node.op, dnn.GpuDnnTransformerGradT)
+                for node in f_gt.maker.fgraph.apply_nodes])
+
+    # Generate random set of RGB images(pixel values in [0, 255]). The set of
+    # images is generated in the expected (NCHW) format
+    input_dims = (5, 3, 16, 16)
+    inputs_val = np.random.randint(low=0, high=256, size=input_dims).astype(theano.config.floatX)
+    # Tensor with transformations
+    theta_val = np.random.random((input_dims[0], 2, 3)).astype(theano.config.floatX)
+
+    # Check that the gradients are computed
+    f_gi(inputs_val, theta_val)
+    f_gt(inputs_val, theta_val)
+
+    def grad_functor(inputs, theta):
+        out = dnn.dnn_spatialtf(inputs, theta)
+        return out
+
+    utt.verify_grad(grad_functor, [inputs_val, theta_val], mode=mode_with_gpu)
```

Comment:
Decrease absolute tolerance to 1e-3 (currently 1e-4) seems to resolve the problem. Can you check it on your computer ? Does this test pass with default absolute tolerance on your computer ?

```python
utt.verify_grad(grad_functor, [inputs_val, theta_val], mode=mode_with_gpu, abs_tol=1e-3)
```

NB: Current spatialtf tests use floatX, which can be float16, float32 or float64, and appropriate tolerance may change from a float type to another (especially with float16). Check on your computer if tests pass with every possible value for floatX.